### PR TITLE
Add more extensions on composer.json, improve phpdocs and remove dead code

### PR DESCRIPTION
**Pull request by tgalopin**
This PR proposes a few things:

- add ctype, iconv and mbstring extensions in composer.json for the same reason as the DOM extension ;
- improve several phpdocs by adding more detailed information on types and arguments
- remove a few unused variables
- fix a few bugs

Don't hesitate to answer my comments when you think I'm wrong :) !

### CODE REVIEW

File: src/HTML5/Parser/DOMTreeBuilder.php [hunk 1]
```diff
@@ -136,6 +136,7 @@ class DOMTreeBuilder implements EventHandler
     protected $stack = array();
 
     protected $current; // Pointer in the tag hierarchy.
+    protected $rules;
```
Comment (tgalopin) [inline src/HTML5/Parser/DOMTreeBuilder.php]: The property was dynamically set in the constructor.

File: src/HTML5/Parser/StringInputStream.php [hunk 2]
```diff
@@ -126,7 +131,7 @@ public function currentLine()
      */
     public function getCurrentLine()
     {
-        return currentLine();
+        return $this->currentLine();
```
Comment (tgalopin) [inline src/HTML5/Parser/StringInputStream.php]: I think this was a bug.

File: src/HTML5/Parser/UTF8Utils.php [hunk 3]
```diff
@@ -108,7 +116,7 @@ public static function convertToUTF8($data, $encoding = 'UTF-8')
             $data = @iconv($encoding, 'UTF-8//IGNORE', $data);
         } else {
             // we can make a conforming native implementation
-            throw new Exception('Not implemented, please install mbstring or iconv');
+            throw new \Exception('Not implemented, please install mbstring or iconv');
```
Comment (tgalopin) [inline src/HTML5/Parser/UTF8Utils.php]: I think this was a bug.

File: src/HTML5/Parser/UTF8Utils.php [hunk 4]
```diff
@@ -124,14 +132,16 @@ public static function convertToUTF8($data, $encoding = 'UTF-8')
...
      * @return array An array of (string) error messages produced by the scanning.
+     *
+     * @throws \Exception When PCRE is not available.
      */
     public static function checkForIllegalCodepoints($data)
     {
-        if (! function_exists('preg_match_all')) {
-            throw\Exception('The PCRE library is not loaded or is not available.');
+        if (!function_exists('preg_match_all')) {
+            throw new \Exception('The PCRE library is not loaded or is not available.');
```
Comment (tgalopin) [inline src/HTML5/Parser/UTF8Utils.php]: I think this was a bug.

File: composer.json [hunk 5]
```diff
@@ -20,8 +20,11 @@
         }
     ],
     "require" : {
+        "ext-ctype": "*",
         "ext-dom": "*",
+        "ext-iconv": "*",
```
Comment (goetas) [inline composer.json]: iconv is used only as fallback when mbstring is not present, so technically both iconv and mbstring are optional (but at least one of them is required)

Any idea on how to express this in composer?

Comment (tgalopin) [reply composer.json]: You are right, I don't think there is any way to ask Composer for that. Let's keep an exception :) .

**Comment by tgalopin** [comment]
Updated :)

**Comment by goetas** [comment]
Thanks! 
Sorry if it took so long to merge it, but somehow I've lost this PR!

**Comment by tgalopin** [comment]
No problem, I didn't update it before anyway :) . Thanks for the merge!

**Pull request merged**